File: Source/NSEntityDescription+Sync.m
#import "NSEntityDescription+Sync.h"

@implementation NSEntityDescription (Sync)

- (nonnull NSArray<NSRelationshipDescription *> *)sync_relationships {
    NSMutableArray *relationships = [NSMutableArray array];

    for (id propertyDescription in [self properties]) {
        if ([propertyDescription isKindOfClass:[NSRelationshipDescription class]]) {
            [relationships addObject:propertyDescription];
        }
    }

    return relationships.copy;
}

- (nullable NSRelationshipDescription *)sync_parentEntity {
    NSArray *relationships = [self sync_relationships];
    NSRelationshipDescription *foundParentEntity = nil;
    for (NSRelationshipDescription *relationship in relationships) {
        BOOL isParent = ([relationship.destinationEntity.name isEqualToString:self.name] &&
                         !relationship.isToMany);
        if (isParent) {
            foundParentEntity = relationship;
        }
    }

    return foundParentEntity;
}

@end

File: Source/NSEntityDescription+Sync.swift
import CoreData

extension NSEntityDescription {
  /**
   Finds the relationships for the current entity.
   - returns The list of relationships for the current entity.
   */
  func sync_relationships() -> [NSRelationshipDescription] {
    var relationships = [NSRelationshipDescription]()

    properties.forEach { propertyDescription in
      if let relationshipDescription = propertyDescription as? NSRelationshipDescription {
        relationships.append(relationshipDescription)
      }
    }

    /**
     Finds the parent for the current entity, if there are many parents nil will be returned.
     - returns The parent relationship for the current entity
     */
    func sync_parentEntity() -> NSRelationshipDescription? {
        let relationships = self.sync_relationships()
        var foundParentEntity: NSRelationshipDescription? = nil
        for relationship in relationships {
            let isParent = relationship.destinationEntity?.name == self.name && !relationship.toMany
            if isParent {
                foundParentEntity = relationship
            }
        }

        return foundParentEntity
    }
}
